scheduler/scheduler.go

```diff
@@ -69,17 +82,113 @@ type Queue struct {
 	queuedRequestsMetricsKey string
 }
 
+// FifoQueue objects implement a FIFO queue for handling requests,
+// with a maximum allowed concurrency and queue size. Currently, they
+// can be used from the fifo filters in the filters/scheduler package
+// only.
+type FifoQueue struct {
+	queue                    *fifoQueue
+	config                   Config
+	metrics                  metrics.Metrics
+	activeRequestsMetricsKey string
+	errorFullMetricsKey      string
+	errorOtherMetricsKey     string
+	errorTimeoutMetricsKey   string
+	queuedRequestsMetricsKey string
+}
+
+type fifoQueue struct {
+	mu             sync.RWMutex
+	counter        *atomic.Uint64
+	sem            *semaphore.Weighted
+	timeout        time.Duration
+	maxQueueSize   uint64
+	maxConcurrency uint64
+	closed         bool
+}
+
+func (fq *fifoQueue) Status() QueueStatus {
+	fq.mu.RLock()
+	defer fq.mu.RUnlock()
+	maxConcurrency := fq.maxConcurrency
+	all := fq.counter.Load()
+	queued := float64(0)
+	if all > maxConcurrency {
+		queued = math.Max(float64(all-maxConcurrency), 0)
```

Comment:
Why do we need math.Max if we know `all > maxConcurrency` ?

Reply:
I saw negative values